File: leetcode/Python/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        sol, cur, m = [], [], ["abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"]
        if not digits:
            return sol
        def _traverse(idx: int):
            if idx==len(digits):
                sol.append("".join(cur))
                return
            for ch in m[ord(digits[idx])-50]:
                cur.append(ch)
                _traverse(idx+1)
                cur.pop()
        _traverse(0)
        return sol

File: leetcode/C++/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> sol;
    vector<string> v = {"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
    string cur;
    void traverse(int idx, string &digits){
        if(idx==digits.length()) {
            sol.push_back(cur);
            return;
        }
        for(char &ch : v[digits[idx]-'2']){
            cur.push_back(ch);
            traverse(idx+1, digits);
            cur.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        if(digits.empty()) return sol;
        traverse(0,digits);
        return sol;
    }
};